validator_client/src/beacon_node_fallback.rs

```diff
@@ -0,0 +1,424 @@
+        let mut n = 0;
+        for candidate in &self.candidates {
+            if candidate.status(RequireSynced::Yes).await.is_ok() {
+                n += 1
+            }
+        }
+        n
+    }
+
+    /// The count of candidates that are online and compatible, but not necessarily synced.
+    pub async fn num_available(&self) -> usize {
+        let mut n = 0;
+        for candidate in &self.candidates {
+            if candidate.status(RequireSynced::No).await.is_ok() {
+                n += 1
+            }
+        }
+        n
+    }
+
+    /// Loop through any `self.candidates` that we don't think are online, compatible or synced and
+    /// poll them to see if their status has changed.
+    ///
+    /// We do not poll nodes that are synced to avoid sending additional requests when everything is
+    /// going smoothly.
+    pub async fn update_unready_candidates(&self) {
+        for candidate in &self.candidates {
+            // There is a potential race condition between having the read lock and the write
+            // lock. The worst case of this race is running `try_become_ready` twice, which is
+            // acceptable.
+            //
+            // Note: `RequireSynced` is always set to false here. This forces us to recheck the sync
+            // status of nodes that were previously not-synced.
+            if candidate.status(RequireSynced::Yes).await.is_err() {
+                // There exists a race-condition that could result in `refresh_status` being called
+                // when the status does not require refreshing anymore. This deemed is an
+                // acceptable inefficiency.
+                let _ = candidate
+                    .refresh_status(self.slot_clock.as_ref(), &self.spec, &self.log)
+                    .await;
+            }
+        }
+    }
+
+    /// Run `func` against each candidate in `self`, returning immediately if a result is found.
+    /// Otherwise, return all the errors encountered along the way.
+    ///
+    /// First this function will try all nodes with a suitable status. If no candidates are suitable
+    /// or all the requests fail, it will try updating the status of all unsuitable nodes and
+    /// re-running `func` again.
+    pub async fn first_success<'a, F, O, Err, R>(
+        &'a self,
+        require_synced: RequireSynced,
+        func: F,
+    ) -> Result<O, AllErrored<Err>>
+    where
+        F: Fn(&'a BeaconNodeHttpClient) -> R,
+        R: Future<Output = Result<O, Err>>,
+    {
+        let mut errors = vec![];
+        let mut to_retry = vec![];
+
+        // First pass: try `func` on all ready candidates.
+        for candidate in &self.candidates {
+            if let Err(e) = candidate.status(require_synced).await {
+                // This client was not ready on the first pass, we might try it again later.
+                to_retry.push(candidate);
+                errors.push((candidate.beacon_node.to_string(), Error::Unavailable(e)));
```

Comment:
Is it wanted that there will be two errors for this candidate in the errors vector? We could change the errors vector a HashMap to only return one error per candidate.

Reply:
I think two errors is fine. We get to see why it failed the first time and then why it failed the second time, this is useful information. E.g., it was offline, then it came online but returned an error.